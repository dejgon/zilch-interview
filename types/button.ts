export enum ButtonType {
  Outlined = 'outlined',
  Contained = 'contained',
}
