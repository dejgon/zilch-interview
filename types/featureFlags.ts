export enum FeatureFlags {
  ApiAddress = 'api_address',
  ApiAddressAndroid = 'api_address_android',
  Maintenance = 'maintenance',
}
