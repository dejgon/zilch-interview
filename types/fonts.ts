export enum Fonts {
  OswaldRegular = 'Oswald-Regular',
  OswaldLight = 'Oswald-Light',
}

export enum FontWeight {
  Normal = '400',
  Bold = '600',
}
