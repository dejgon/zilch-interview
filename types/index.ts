export * from './button';
export * from './fonts';
export * from './rootStackParamList';
export * from './routes';
export * from './secureStoreKeys';
export * from './user';
export * from './featureFlags';
export * from './requestStatus';
export * from './person';
