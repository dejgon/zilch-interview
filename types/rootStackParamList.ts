import { Person } from './person';

export type RootStackParamList = {
  Home: undefined;
  SignIn: undefined;
  Swapi: undefined;
  SwapiCharacter: {
    person: Person;
  };
  Dashboard: undefined;
  Maintenance: undefined;
};
