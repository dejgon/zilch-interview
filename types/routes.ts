export enum Routes {
  Home = 'Home',
  SignIn = 'SignIn',
  Swapi = 'Swapi',
  SwapiCharacter = 'SwapiCharacter',
  Dashboard = 'Dashboard',
  Maintenance = 'Maintenance',
}
