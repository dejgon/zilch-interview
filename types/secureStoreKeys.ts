export enum SecureStoreKeys {
  User = 'user',
}
