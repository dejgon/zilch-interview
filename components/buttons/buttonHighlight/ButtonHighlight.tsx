import React from 'react';
import { Text, TouchableOpacity } from 'react-native';

import useStyles from './styles';
import { Colors } from '../../../consts';
import { ButtonType } from '../../../types';

interface Props {
  text: string;
  type: ButtonType;
  onPress: () => void;
  isDisabled?: boolean;
}

export function ButtonHighlight(props: Props) {
  const { text, type, onPress, isDisabled } = props;

  const { button, buttonText, disabled, contained, outlined } = useStyles;

  const getStyleBasedOnType = () => {
    return type === ButtonType.Outlined ? outlined : contained;
  };

  const getTextColorBasedOnType = () => {
    return type === ButtonType.Outlined
      ? Colors.Green.GREEN_500
      : Colors.Fonts.WHITE;
  };

  return (
    <TouchableOpacity
      testID="button-highlight"
      onPress={onPress}
      style={{
        ...button,
        ...getStyleBasedOnType(),
        ...(isDisabled && disabled),
      }}
      disabled={isDisabled}>
      <Text style={{ ...buttonText, color: getTextColorBasedOnType() }}>
        {text}
      </Text>
    </TouchableOpacity>
  );
}
