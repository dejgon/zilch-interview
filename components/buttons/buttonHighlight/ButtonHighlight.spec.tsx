import React from 'react';

import { act, fireEvent, render } from '@testing-library/react-native';

import { ButtonHighlight } from './ButtonHighlight';
import { ButtonType } from '../../../types';

describe('ButtonHighlight.tsx', () => {
  let onPress: jest.Mock;

  beforeEach(() => {
    onPress = jest.fn();
  });

  it('renders contained', () => {
    const component = render(
      <ButtonHighlight
        type={ButtonType.Contained}
        text="Contained button"
        onPress={onPress}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders outlined', () => {
    const component = render(
      <ButtonHighlight
        type={ButtonType.Outlined}
        text="Outlined button"
        onPress={onPress}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('triggers onPress properly', () => {
    const { getByTestId } = render(
      <ButtonHighlight
        type={ButtonType.Contained}
        text="Contained button"
        onPress={onPress}
      />
    );

    act(() => {
      fireEvent.press(getByTestId('button-highlight'));
    });

    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it('renders disabled', () => {
    const { getByTestId } = render(
      <ButtonHighlight
        isDisabled
        type={ButtonType.Contained}
        text="Disabled button"
        onPress={onPress}
      />
    );

    const button = getByTestId('button-highlight');

    act(() => {
      fireEvent.press(button);
    });

    expect(button.props.disabled).toBeFalsy();
    expect(onPress).not.toHaveBeenCalled();
  });
});
