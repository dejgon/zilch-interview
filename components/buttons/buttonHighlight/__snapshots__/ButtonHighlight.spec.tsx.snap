// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonHighlight.tsx renders contained 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007FFF",
      "borderRadius": 10,
      "borderWidth": 0,
      "color": "#FFFFFF",
      "display": "flex",
      "flexDirection": "row",
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "width": 326,
    }
  }
  testID="button-highlight"
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "Oswald-Regular",
        "fontSize": 24,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 36,
      }
    }
  >
    Contained button
  </Text>
</View>
`;

exports[`ButtonHighlight.tsx renders outlined 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#007FFF",
      "borderRadius": 10,
      "borderStyle": "solid",
      "borderWidth": 2,
      "color": "#007FFF",
      "display": "flex",
      "flexDirection": "row",
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "width": 326,
    }
  }
  testID="button-highlight"
>
  <Text
    style={
      {
        "color": "#007FFF",
        "fontFamily": "Oswald-Regular",
        "fontSize": 24,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 36,
      }
    }
  >
    Outlined button
  </Text>
</View>
`;

exports[`ButtonHighlight.tsx should render basic snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007FFF",
      "borderRadius": 10,
      "borderWidth": 0,
      "color": "#FFFFFF",
      "display": "flex",
      "flexDirection": "row",
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "width": 326,
    }
  }
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "Oswald-Regular",
        "fontSize": 24,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 36,
      }
    }
  >
    Contained button
  </Text>
</View>
`;
